deeplake/core/transform/transform_dataset.py

```diff
@@ -1,40 +1,204 @@
+                (LinkedSample, Tensor, type(None), PartialSample, LinkedTiledSample),
+            ):
+                values.append(np.asarray(item))
+            else:
+                values.append(item)
+        if squeeze:
+            values = values[0]
+        return values
+
+    def numpy_compressed(self):
+        idx = self.idx
+        if self.numpy_only:
+            if isinstance(idx, int):
+                i = bisect.bisect_right(self.cum_sizes, idx)
+                if i == 0:
+                    j = idx
+                else:
+                    j = idx - self.cum_sizes[i - 1]
+                return self.items[i][j]
+        return self.items[idx]
+
+    def non_numpy_only(self):
+        if self.numpy_only:
+            items = list(chain(*self.items[:]))
+            self.items.clear()
+            self.items += items
+            self.cum_sizes.clear()
+            self.numpy_only = False
+
+    def append(self, item):
+        if self.is_group:
+            raise TensorDoesNotExistError(self.name)
+        if self.numpy_only:
+            # optimization applicable only if extending
+            self.non_numpy_only()
+        self.items.append(item)
+        if self.dataset.all_chunk_engines:
+            self.dataset.item_added(item)
+
+    def extend(self, items):
+        if self.numpy_only:
+            if isinstance(items, np.ndarray):
+                self.items.append(items)
+                if len(self.cum_sizes) == 0:
+                    self.cum_sizes.append(len(items))
+                else:
+                    self.cum_sizes.append(self.cum_sizes[-1] + len(items))
+                if self.dataset.all_chunk_engines:
+                    self.dataset.item_added(items)
+                return
+            else:
+                self.non_numpy_only()
+
+        for item in items:
+            self.append(item)
 
 
 class TransformDataset:
-    def __init__(self, all_tensors=None, slice_list=None):
-        """Creates a Dataset like object that supports "." access of tensors and appends/extends to the tensors.
-        This is used as sample_out in deeplake transforms.
-        """
-        self.tensors = all_tensors or {}
-        self.slice_list = slice_list or []
+    def __init__(
+        self,
+        tensors,
+        all_chunk_engines=None,
+        group_index="",
+        label_temp_tensors=None,
+        idx=slice(None, None, None),
+        cache_size=16,
+    ):
+        self.tensors = tensors
+        self.data = {tensor: TransformTensor(self, tensor) for tensor in tensors}
+        self.all_chunk_engines = all_chunk_engines
+        self.group_index = group_index
+        self.label_temp_tensors = label_temp_tensors
+        self.cache_size = cache_size * MB
+        self.cache_used = 0
+        self.idx = idx
+        self.pg_callback = None
 
     def __len__(self):
-        return min(len(self[tensor]) for tensor in self.tensors)
+        return max(len(self[tensor]) for tensor in self.data)
```

Comment:
why max?

Reply:
this is better in case length of tensors are not same, all samples will still be accessible